crates/stages/src/stages/headers.rs

```diff
@@ -218,6 +221,61 @@ where
 
         info!(target: "sync::stages::headers", len = downloaded_headers.len(), "Received headers");
 
+        let tip_block_number = match tip {
+            // If tip is hash and it equals to the first downloaded header's hash, we can use
+            // the block number of this header as tip.
+            BlockHashOrNumber::Hash(hash) => downloaded_headers.first().and_then(|header| {
+                if header.hash == hash {
+                    Some(header.number)
+                } else {
+                    None
+                }
+            }),
+            // If tip is number, we can just grab it and not resolve using downloaded headers.
+            BlockHashOrNumber::Number(number) => Some(number),
+        };
+
+        // Since we're syncing headers in batches, gap tip will move in reverse direction towards
+        // our local head with every iteration. To get the actual target block number we're
+        // syncing towards, we need to take into account already synced headers from the database.
+        // It is `None`, if tip didn't change and we're still downloading headers for previously
+        // calculated gap.
+        let target_block_number = if let Some(tip_block_number) = tip_block_number {
+            let local_max_block_number = tx
+                .cursor_read::<tables::CanonicalHeaders>()?
+                .last()?
+                .map(|(canonical_block, _)| canonical_block);
+
+            Some(tip_block_number.max(local_max_block_number.unwrap_or(tip_block_number)))
+        } else {
+            None
+        };
+
+        let mut stage_checkpoint = current_progress
+            .headers_stage_checkpoint()
```

Comment:
however, this makes it handling checkpoints inside the change less convenient, even though we now what kind of checkpoint we have